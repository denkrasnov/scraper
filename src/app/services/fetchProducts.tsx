import React, {
  useEffect,
  useReducer,
  useContext,
  createContext,
  Dispatch,
  FC
} from "react";
import axios from "axios";

import {
  Action,
  State,
  ActionTypes,
  ProductsProviderProps,
  Product
} from "./types";

const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case ActionTypes.FETCH:
      return {
        ...state,
        query: action.payload,
        isLoading: true,
        isError: ""
      };
    case ActionTypes.FETCH_INIT:
      return {
        ...state,
        isLoading: true,
        isError: ""
      };
    case ActionTypes.FETCH_ERROR:
      return {
        ...state,
        isLoading: false,
        isError: action.payload
      };
    case ActionTypes.FETCH_SUCCESS:
      return {
        ...state,
        isLoading: false,
        products: action.payload
      };
    default:
      throw new Error();
  }
};

const initialState = {
  isLoading: false,
  isError: "",
  query: "",
  products: null
};

const fetchProducts = (): [State, Dispatch<Action>] => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const { query } = state;

  useEffect(() => {
    let didCancel = false;

    const fetchData = async () => {
      dispatch({ type: ActionTypes.FETCH_INIT });
      try {
        const result: {
          data: { products: Product[] | null };
        } = await axios.post("/search", { query }); // TODO": GET and type data

        if (!didCancel) {
          dispatch({
            type: ActionTypes.FETCH_SUCCESS,
            payload: result.data.products
          });
        }
      } catch (error) {
        if (!didCancel) {
          dispatch({ type: ActionTypes.FETCH_ERROR, payload: error });
        }
        // eslint-disable-next-line no-console
        console.error("Something went wrong:", error);
      }
    };

    if (query) {
      fetchData();
    }

    return () => {
      didCancel = true;
    };
  }, [query]);

  return [state, dispatch];
};

interface MainReducer {
  products: [State, Dispatch<Action>];
}

export const StateContext = createContext<MainReducer>({
  products: [initialState, () => {}]
});

const mainReducer = () => ({
  products: fetchProducts()
});

export const ProductsProvider: FC<ProductsProviderProps> = ({ children }) => (
  <StateContext.Provider value={mainReducer()}>
    {children}
  </StateContext.Provider>
);

export const useFullContext = () => useContext(StateContext);

export default ProductsProvider;
