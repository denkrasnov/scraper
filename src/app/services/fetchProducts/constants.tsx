export const INITIAL_STATE = {
  isLoading: false,
  isError: "",
  query: "",
  products: null
};
