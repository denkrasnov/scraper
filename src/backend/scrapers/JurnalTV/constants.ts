export const JurnalTV = "https://jurnaltv.md/page/ultima-ora/1";
