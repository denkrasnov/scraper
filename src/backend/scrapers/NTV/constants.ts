export const NTVmd = "http://ntv.md/index.php?do=cat&category=lentamd";
export const NTVru = "http://ntv.md/ru/index.php?do=cat&category=lenta";
