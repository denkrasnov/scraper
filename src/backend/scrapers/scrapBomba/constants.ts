const tvs =
  "https://bomba.md/ru/category/633647/?price%5Bmin%5D=1299&price%5Bmax%5D=69999&filters%5B2817%5D%5Bmin%5D=22&filters%5B2817%5D%5Bmax%5D=75";

const fridges = [
  "https://bomba.md/ru/category/tehnica-de-bucatarie-electrocasnice-mari-frigidere-incorporabile/?price%5Bmin%5D=7499&price%5Bmax%5D=54999&filters%5B2256%5D%5Bmin%5D=178&filters%5B2256%5D%5Bmax%5D=311&filters%5B2250%5D%5Bmin%5D=41&filters%5B2250%5D%5Bmax%5D=122&filters%5B2219%5D%5Bmin%5D=219&filters%5B2219%5D%5Bmax%5D=490&filters%5B1485%5D%5Bmin%5D=1&filters%5B1485%5D%5Bmax%5D=2&filters%5B720%5D%5Bmin%5D=144&filters%5B720%5D%5Bmax%5D=194&filters%5B4055%5D%5Bmin%5D=54&filters%5B4055%5D%5Bmax%5D=112&filters%5B786%5D%5Bmin%5D=54&filters%5B786%5D%5Bmax%5D=55",
  "https://bomba.md/ru/category/tehnica-de-bucatarie-electrocasnice-mari-frigidere-cu-congelator-jos/?price%5Bmin%5D=3699&price%5Bmax%5D=44999&filters%5B720%5D%5Bmin%5D=150&filters%5B720%5D%5Bmax%5D=203&filters%5B4055%5D%5Bmin%5D=55&filters%5B4055%5D%5Bmax%5D=70&filters%5B786%5D%5Bmin%5D=60&filters%5B786%5D%5Bmax%5D=80&filters%5B2219%5D%5Bmin%5D=243&filters%5B2219%5D%5Bmax%5D=505&filters%5B2256%5D%5Bmin%5D=160&filters%5B2256%5D%5Bmax%5D=400&filters%5B2250%5D%5Bmin%5D=54&filters%5B2250%5D%5Bmax%5D=113",
  "https://bomba.md/ru/category/tehnica-de-bucatarie-electrocasnice-mari-frigidere-cu-congelator-sus/?price%5Bmin%5D=2799&price%5Bmax%5D=17999&filters%5B720%5D%5Bmin%5D=106&filters%5B720%5D%5Bmax%5D=192&filters%5B4055%5D%5Bmin%5D=47&filters%5B4055%5D%5Bmax%5D=83&filters%5B786%5D%5Bmin%5D=50&filters%5B786%5D%5Bmax%5D=77&filters%5B2219%5D%5Bmin%5D=110&filters%5B2219%5D%5Bmax%5D=618&filters%5B2256%5D%5Bmin%5D=80&filters%5B2256%5D%5Bmax%5D=459&filters%5B2250%5D%5Bmin%5D=30&filters%5B2250%5D%5Bmax%5D=170",
  "https://bomba.md/ru/category/tehnica-de-bucatarie-electrocasnice-mari-frigidere-side-by-side/?price%5Bmin%5D=9999&price%5Bmax%5D=29999&filters%5B720%5D%5Bmin%5D=177&filters%5B720%5D%5Bmax%5D=192&filters%5B4055%5D%5Bmin%5D=80&filters%5B4055%5D%5Bmax%5D=92&filters%5B786%5D%5Bmin%5D=64&filters%5B786%5D%5Bmax%5D=75&filters%5B2219%5D%5Bmin%5D=455&filters%5B2219%5D%5Bmax%5D=647&filters%5B2256%5D%5Bmin%5D=269&filters%5B2256%5D%5Bmax%5D=418&filters%5B2250%5D%5Bmin%5D=102&filters%5B2250%5D%5Bmax%5D=236&filters%5B1485%5D%5Bmin%5D=2&filters%5B1485%5D%5Bmax%5D=4",
  "https://bomba.md/ru/category/tehnica-de-bucatarie-electrocasnice-mari-frigidere-cu-o-usa/?price%5Bmin%5D=1799&price%5Bmax%5D=8899&filters%5B2256%5D%5Bmin%5D=39&filters%5B2256%5D%5Bmax%5D=341&filters%5B2250%5D%5Bmin%5D=0&filters%5B2250%5D%5Bmax%5D=14&filters%5B2219%5D%5Bmin%5D=0&filters%5B2219%5D%5Bmax%5D=341&filters%5B720%5D%5Bmin%5D=47&filters%5B720%5D%5Bmax%5D=190&filters%5B4055%5D%5Bmin%5D=45&filters%5B4055%5D%5Bmax%5D=60&filters%5B786%5D%5Bmin%5D=45&filters%5B786%5D%5Bmax%5D=64"
];

export const urls = [tvs, ...fridges];
