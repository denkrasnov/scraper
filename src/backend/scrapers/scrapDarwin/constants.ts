const tvs = "https://darwin.md/ru/tv-foto/tv";

const fridges =
  "https://darwin.md/ru/electrocasnice/electrocasnice-mari/frigidere";

export const urls = [tvs, fridges];
