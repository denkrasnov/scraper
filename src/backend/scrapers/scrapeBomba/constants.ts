import { ProductName } from "../types";

const tv =
  "https://bomba.md/ru/category/633647/?price%5Bmin%5D=1299&price%5Bmax%5D=69999&filters%5B2817%5D%5Bmin%5D=22&filters%5B2817%5D%5Bmax%5D=75";

const fridge =
  "https://bomba.md/ru/category/tehnica-de-bucatarie-electrocasnice-mari-frigidere/?price%5Bmin%5D=1699&price%5Bmax%5D=54999&filters%5B720%5D%5Bmin%5D=47&filters%5B720%5D%5Bmax%5D=203&filters%5B4055%5D%5Bmin%5D=45&filters%5B4055%5D%5Bmax%5D=112&filters%5B2219%5D%5Bmin%5D=0&filters%5B2219%5D%5Bmax%5D=647&filters%5B2256%5D%5Bmin%5D=39&filters%5B2256%5D%5Bmax%5D=459&filters%5B2250%5D%5Bmin%5D=0&filters%5B2250%5D%5Bmax%5D=236&filters%5B3736%5D%5Bmin%5D=0&filters%5B3736%5D%5Bmax%5D=45&filters%5B1496%5D%5Bmin%5D=1&filters%5B1496%5D%5Bmax%5D=2";

export const urls = [
  { url: tv, name: ProductName.TV },
  { url: fridge, name: ProductName.FRIDGE }
];
