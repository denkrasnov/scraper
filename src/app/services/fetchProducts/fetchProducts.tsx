import { useEffect, useReducer, Dispatch } from "react";
import axios from "axios";

import { INITIAL_STATE } from "./constants";
import reducer from "./reducer";
import { Action, State, ActionTypes, Product } from "./types";

const fetchProducts = (): [State, Dispatch<Action>] => {
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
  const { query } = state;

  useEffect(() => {
    let didCancel = false;

    const fetchData = async () => {
      dispatch({ type: ActionTypes.FETCH_INIT });

      try {
        const result: {
          data: { products: Product[] | null };
        } = await axios.post("/search", { query }); // TODO": GET and type data

        if (!didCancel) {
          dispatch({
            type: ActionTypes.FETCH_SUCCESS,
            payload: result.data.products
          });
        }
      } catch (error) {
        if (!didCancel) {
          dispatch({ type: ActionTypes.FETCH_ERROR, payload: error });
        }
        // eslint-disable-next-line no-console
        console.log("Something went wrong:", error);
      }
    };

    if (query) {
      fetchData();
    }

    return () => {
      didCancel = true;
    };
  }, [query]);

  return [state, dispatch];
};

export default fetchProducts;
