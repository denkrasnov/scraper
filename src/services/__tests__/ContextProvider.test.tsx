import React from "react";
import { shallow } from "enzyme";
import { renderHook, cleanup } from "@testing-library/react-hooks";

import { INITIAL_STATE } from "../fetchProducts/constants";
import ContextProvider, { useFullContext } from "../ContextProvider";

describe("ContextProvider", () => {
  afterEach(() => {
    cleanup();
  });

  it("should render", () => {
    const children = <div />;
    const component = shallow(<ContextProvider>{children}</ContextProvider>);

    expect(component).toMatchSnapshot();
  });

  describe("useFullContext", () => {
    it("should return", () => {
      const { result } = renderHook(() => useFullContext());
      expect(result.current).toEqual(
        expect.arrayContaining([INITIAL_STATE, expect.any(Function)])
      );
    });
  });
});
