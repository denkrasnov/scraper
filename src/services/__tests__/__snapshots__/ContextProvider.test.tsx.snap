// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextProvider should render 1`] = `
<ContextProvider
  value={
    Array [
      Object {
        "isError": "",
        "isLoading": false,
        "products": null,
        "query": "",
      },
      [Function],
    ]
  }
>
  <div />
</ContextProvider>
`;
